### ISSUE TITLE
Support for running in shadow DOM

### ISSUE THREAD

**Issue by Reinmar** [ckeditor/ckeditor5#3891]
Apparently, shadow DOM has its own selection (like an iframe), so when running in a shadow root, the engine must use that selection.

Welcome to hell...

**Comment by ywsang** [comment]
@Reinmar Hi there! I’ve been working on creating a web component version of CKEditor5, which involves putting the editor inside of shadow DOM. I’ve made a couple changes to the source code that help increase the amount of support for running the editor in the shadow DOM. So far, these changes are related to the editor’s styles (so that the editor looks visually correct when placed inside shadow DOM) and the engine selection issue. I’d love to contribute these changes back to the repo, though they’re not complete fixes for supporting running the editor in shadow DOM (this issue). Would you be open to accepting incremental changes towards this goal that tackle more fine-grained issues? 

cc: @justinfagnani

**Comment by yoyo837** [comment]
Can we consider this in 2022?

**Comment by ralfpatric** [comment]
Any idea if this will come in 2023? 🤔

**Comment by Witoso** [comment]
@ralfpatric most likely not, we have not currently set it as a priority.

**Comment by mancristiana** [comment]
I reproduced the issue we are facing when adding CKEditor in ShadowDOM
https://stackblitz.com/edit/lit-element-ckeditor5-xd7j6g?file=index.html,index.js,package.json

**Comment by Reinmar** [comment]
I don't want to be the party killer here but the problem with shadow DOM isn't (to our knowledge) really fixable at the moment. It's been a while since we investigated this and AFAICS we failed to share the details here, but all issues except one were rather straightforward to fix.

The one issue being: selection handling 😳🤬

There's currently no standardized API to access selection in a shadow root. If there were even different APIs, but somehow compatible, we would painfully resolve this ticket. But nope – it's 2023 and the state is as follows:

*   Chrome: non-standard API (`shadowRoot#getSelection()`)
*   Webkit: [very fresh implementation](https://bugs.webkit.org/show_bug.cgi?id=163921) based on an incomplete/unreviewed specification
*   Firefox: [https://bugzilla.mozilla.org/show_bug.cgi?id=1430308](https://bugzilla.mozilla.org/show_bug.cgi?id=1430308) "The specification isn't quite ready yet, so I don't know what they have shipped in their beta"

When researching this over a year ago we checked crazy hacks that we found. AFAIR, those were in line with: [https://github.com/jsturgill/shadow-root-get-selection-polyfill](https://github.com/jsturgill/shadow-root-get-selection-polyfill.). Unfortunately, our QA team found many instabilities of those hacks. I can share the results if someone's interested, but it really didn't look promising.

The thing is – all editor developers are struggling with this unless they are using an `<iframe>` to wrap the editable element. And while we could ship some shadow DOM support with those hacks, that'd be a false promise of a usable feature.

I [pinged Firefox guys](https://bugzilla.mozilla.org/show_bug.cgi?id=1430308#c6) because I can't believe we'll see the spec for this in the next 3 years. If there will be no reaction, I think that we'll consider going with no hacks but also no Firefox support (or hacks to support Firefox, but without aiming at making it super stable). This is depressing but it may be the only choice.

**Comment by 2dareis2do** [comment]
With regards

> There's currently no standardized API to access selection in a shadow root. If there were even different APIs, but somehow compatible, we could somehow resolve this ticket. But nope – it's 2023 and the state is as follows:

Is it possible to use some form of polyfill? Maybe something like this:

https://github.com/GoogleChromeLabs/shadow-selection-polyfill

**Comment by Witoso** [comment]
@2dareis2do, thanks for your suggestions! Tiny under the hood is still a wrapper on an iframe, that's why the issues mentioned in this thread are not relevant to them.

![](https://github.com/ckeditor/strapi-plugin-ckeditor/assets/9881379/d034ca08-2f8d-42a2-8111-6e835c321587)

**Comment by jogibear9988** [comment]
As it looks with the new API from Safari, ShadowDom could now be supported.
At least for me it works in my Webcomponent Designer Application with this Code: 
https://github.com/node-projects/web-component-designer/blob/master/packages/web-component-designer/src/elements/helper/SelectionHelper.ts

**Comment by Witoso** [comment]
Yes, v17 or 17.1 introduced `getComposedRanges` as @Reinmar stated, most likely it stabilized a bit in the latest four versions. FF is the biggest blocker for us right now, but AFAICS they haven't replied to our ping :face\_exhaling:

**Comment by jogibear9988** [comment]
But FF works with the document API,

**Comment by hsinyi** [comment]
FWIW,  there are activities and patches in [this Firefox ShadowDOM selection issue ](https://bugzilla.mozilla.org/show_bug.cgi?id=1867058).

**Comment by jogibear9988** [comment]
In my designer (https://node-projects.github.io/web-component-designer-demo/index.html) TextEditing in ShadowDom works with the current Firefox Version using the document.getSelection API

**Comment by Witoso** [comment]
I understand, and congrats on your project, but we would like to minimize the special handling around selection to a minimum. I will keep in mind to test the proposed solution, and it's also reassuring that ProseMirror handles it this way. But it will likely not happen before other projects related to installation (#15502).

**Comment by Joel-Levi** [comment]
With #15502 being closed, is there any update you can share on this issue?

**Comment by Witoso** [comment]
We are conducting some internal tests to understand better the current state of the ecosystem in #16975.

**Comment by itsmebasti** [comment]
Hi there, almost a year passed by since the last update, and we are currently looking for n editor inside out lit app. What's the status here, will this be continued in the near future or do we have to consider it failed.

**Comment by Witoso** [comment]
Yes, we plan to pick it up sooner rather than later, as we are discussing with @Reinmar next year's roadmap. @niegowski's test has proven quite fruitful. We still have some questions on the editor's side, but the ecosystem looks ready.

**Comment by niegowski** [comment]
**Update: shadow DOM and web component support has landed — available for testing in the nightly builds.**

CKEditor 5 now runs inside a shadow root, and inside a `<slot>` of your own web component. Selection, focus, positioning, scrolling and the floating UI layer — balloons, dialogs, tooltips — work there. The editor's floating UI is also no longer bound to `document.body`: the new [`config.ui.overlayContainer`](https://ckeditor5.github.io/docs/nightly/ckeditor5/latest/api/module_core_editor_editorconfig-UiConfig.html#member-overlayContainer) lets you say which element or shadow root hosts it. Set it whenever the editor runs inside a shadow root.

One thing to know before you try it: a shadow root is a separate styling boundary, so the editor stylesheets have to be loaded into every root that holds editor UI, including the overlay container. CSS variables are now declared on both `:root` and `:host`, which means overriding one on `:root` no longer reaches an editor inside a shadow root — override it on the shadow host element instead. The [Styles inside a shadow DOM](https://ckeditor5.github.io/docs/nightly/ckeditor5/latest/getting-started/setup/css.html#styles-inside-a-shadow-dom) section of the CSS guide covers it.

**We would really like your feedback while this is still in nightly.** Does it work in your setup, and with which framework or component library? Are you creating the editor inside a shadow root you own, or slotting it into a component of your own? And is there anything in your integration that still misbehaves?

**Comment by Joel-Levi** [comment]
I refactored our editor component, which used to run CK in an iframe to get around this shadow DOM issue. We have a pretty involved setup were we use our own toolbar (which was outside of that iframe). We also hook into paste pipelines and do a bunch of things with the data down- and upcasts. We also have a lot of test that check where focus goes and where the caret ends up after certain actions. This all to give you an indication of what we test. 

We do not use any floating CK UI.
The stuff around our CK integration is written in Lit.
Using 0.0.0-nightly-20260920.0

All our 700+ playwright tests succeed in chromium. Webkit and Firefox runs have some failures, but I am not seeing the same problems reproducing the tests by hand in the browsers, so probably something else going on there on my end.

I think this is great work!
